poly/src/domain/radix2/fft.rs

```diff
@@ -137,109 +137,131 @@ impl<F: FftField> Radix2EvaluationDomain<F> {
+
+            // Only compact roots to achieve cache locality/compactness if
+            // the roots lookup is done a significant amount of times
+            // Which also implies a large lookup stride.
+            if num_chunks >= MIN_COMPACTION_CHUNKS {
+                if !first {
+                    roots = cfg_into_iter!(roots).step_by(step * 2).collect()
+                }
+                step = 1;
+                roots.shrink_to_fit();
+            } else {
+                step = num_chunks;
+            }
+            first = false;
 
-            ark_std::cfg_chunks_mut!(xi, chunk_size).for_each(|cxi| {
+            cfg_chunks_mut!(xi, chunk_size).for_each(|cxi| {
                 let (lo, hi) = cxi.split_at_mut(gap);
                 // If the chunk is sufficiently big that parallelism helps,
                 // we parallelize the butterfly operation within the chunk.
-                //
-                // if chunk_size > MIN_CHUNK_SIZE_FOR_PARALLELIZATION
-                if gap > MIN_CHUNK_SIZE_FOR_PARALLELIZATION / 2 {
-                    cfg_iter_mut!(lo).zip(hi).enumerate().for_each(butterfly_fn);
+
+                if gap > MIN_PROBLEM_SIZE && num_chunks < max_threads {
+                    cfg_iter_mut!(lo)
+                        .zip(cfg_iter_mut!(hi))
+                        .zip(cfg_iter!(roots).step_by(step))
+                        .for_each(Self::butterfly_fn_io);
                 } else {
-                    lo.iter_mut().zip(hi).enumerate().for_each(butterfly_fn);
+                    lo.iter_mut()
+                        .zip(hi)
+                        .zip(roots.iter().step_by(step))
+                        .for_each(Self::butterfly_fn_io);
                 }
             });
 
-            // Cache align the FFT roots in the sequential case.
-            // In this case, we are aiming to make every root that is accessed one after another,
-            // appear one after another in the list of roots.
-            // if the roots are cache aligned in iteration i, then in iteration i+1,
-            // cache alignment requires selecting every other root.
-            // (The even powers relative to the current iterations generator)
-            //
-            //
-            // (Roots are already aligned in the first iteration,
-            // so we only to do realignment after the first iteration.)
-            #[cfg(not(feature = "parallel"))]
-            {
-                for i in 1..(root_len / 2) {
-                    roots[i] = roots[i * 2];
-                }
-                root_len /= 2;
-            }
             gap /= 2;
         }
     }
 
     fn oi_helper<T: DomainCoeff<F>>(&self, xi: &mut [T], root: F) {
-        let roots = self.roots_of_unity(root);
+        let roots_cache = self.roots_of_unity(root);
+        let compaction_size = core::cmp::min(
+            roots_cache.len() / 2,
+            roots_cache.len() / MIN_COMPACTION_CHUNKS,
+        );
```

Comment:
What happens when `roots_cache.len()` is `2` or less than `MIN_COMPACTION_CHUNKS`? Could you amend the tests to check that as well? Thanks!

Reply:
Hmm the compaction wouldn't happen. So we don't have to worry about it. Notice that min compare is only necessary for MIN_COMPACTION_CHUNKS = 1, since chunks > 0.

Reply:
Ok sounds good, a comment to that effect would be great.